Add aggregate_list utility function, along with unit tests

mgduda opened the pull request:
This PR adds a new utility function, `aggregate_list`, that may be used to
gather arbitrarily sized lists of `SMIOL_Offset` values from all ranks in
a communicator onto a chosen root task in the communicator.

MiCurry commented on src/smiol_utils.c:
```diff
@@ -368,6 +368,126 @@ int transfer_field(const struct SMIOL_decomp *decomp, int dir,
+	if (err != MPI_SUCCESS) {
+		fprintf(stderr, "Error: MPI_Gather failed in aggregate_list\n");
+		return 1;
+	}
```
This `MPI_Gather`, and the `MPI_Reduce` on lines 436-437 seem to be doing relatively the same task, should we reduce them into one? e.g.:

``` C
...

// Allocate counts and displs using comm_size
if (comm_rank == root) {
	*counts = (int *)malloc(sizeof(int) * (size_t)(comm_size));
	*displs = (int *)malloc(sizeof(int) * (size_t)(comm_size));
}

// Gather input counts
err = MPI_Gather((const void *)&n_in32, 1, MPI_INT,
	                      (void *)(*counts), 1, MPI_INT, root, comm);

// Determine n_out
if (comm_rank == root) {
    n_out = 0;
    for (i = 0 ; i < comm_size ; i++) {
        n_out += counts[i];
    } 
}

// Allocate out_list using n_out
if (comm_rank == root) {
	*out_list = (SMIOL_Offset *)malloc(sizeof(SMIOL_Offset)
		                                  * (*n_out))
}

...
```

MiCurry replied:
We could also combine the loop on 467 with the one I proposed above.

mgduda replied:
Those are good suggestions. I'll push changes to remove the `MPI_Reduce` and to combine loops for computing `displs` and `n_out`.

mgduda commented on src/smiol_utils.c:
```diff
@@ -368,6 +368,126 @@ int transfer_field(const struct SMIOL_decomp *decomp, int dir,
+	int err;
+	int i;
+	int n_in32;
+	int n_out32;
```
The name was simply to differentiate these variables, though the 32 was definitely a suggestion that these are probably 32-bit values (though the code in no way depends on this). I'll rename `n_in32` to `n_in_i`.

mgduda commented:
@MiCurry I've just pushed an update to this branch with changes based on your suggestions.

MiCurry commented on smiol_runner.c:
```diff
@@ -3796,6 +3808,400 @@ int test_io_decomp(FILE *test_log)
+	/* Aggregate a list across a communicator of size 1 (using MPI_COMM_SELF) */
+	fprintf(test_log, "Testing aggregation of lists across communicator of size 1: ");
+	n_in = 1000;
+	n_out = -1;
```
The PGI throws the following warning on this line (and the ones below): 

``` c
"smiol_runner.c", line 3848: warning: integer conversion resulted in a change
          of sign
        n_out = -1;
```

Setting an unsigned integer to -1 seems pretty common and not necessarily against C standard, from what I can tell. GNU and Intel seem to be okay with it. Regardless, we can get rid of these warnings by using the `SIZE_MAX` macro from `stdint.h`:

```c
n_out = SIZE_MAX;
```

mgduda replied:
Fixed.

MiCurry commented on smiol_runner.c:
```diff
@@ -3796,6 +3808,400 @@ int test_io_decomp(FILE *test_log)
+	/* Aggregate zero-sized lists */
+	fprintf(test_log, "Testing aggregation of zero-sized lists on all tasks: ");
+	n_in = 0;
+	n_out = -1;
```
``` c
n_out = SIZE_MAX;
```

mgduda replied:
Fixed.

MiCurry commented on smiol_runner.c:
```diff
@@ -3796,6 +3808,400 @@ int test_io_decomp(FILE *test_log)
+		/* Aggregate one-sized lists */
+		fprintf(test_log, "Testing aggregation to %i of one-sized lists on all tasks: ", root);
+		n_in = 1;
+		n_out = -1;
```
PGI warning appears here as well..

mgduda replied:
Fixed.

MiCurry commented on smiol_runner.c:
```diff
@@ -3796,6 +3808,400 @@ int test_io_decomp(FILE *test_log)
+		/* Aggregate N-sized lists */
+		fprintf(test_log, "Testing aggregation to %i of N-sized lists on all tasks: ", root);
+		n_in = 1000;
+		n_out = -1;
```
PGI warning appears here as well..

mgduda replied:
Fixed.

MiCurry commented on smiol_runner.c:
```diff
@@ -3796,6 +3808,400 @@ int test_io_decomp(FILE *test_log)
+		} else {
+			n_in = 10 * (world_rank / 2);
+		}
+		n_out = -1;
```
PGI warning appears here as well..

mgduda replied:
Fixed.

MiCurry commented:
Other than the PGI warnings this branch looks good to go 👍

mgduda commented:
@MiCurry I've just pushed an additional commit (which I'll squash before merging this PR, though for now it seemed easier to see what I've changed in a separate commit) that addresses the PGI warnings.

Additionally, I've addressed these two warnings in `smiol_utils.c` that were generated by the clang compiler:
```
smiol_utils.c:452:12: warning: implicit conversion changes signedness: 'int' to 'size_t' (aka 'unsigned long') [-Wsign-conversion]
                *n_out = (*counts)[0];
                       ~ ^~~~~~~~~~~~
smiol_utils.c:455:14: warning: implicit conversion changes signedness: 'int' to 'unsigned long' [-Wsign-conversion]
                        *n_out += (*counts)[i];
                               ~~ ^~~~~~~~~~~~
```
In the `smiol_runner.c` test code, I also addressed a substantial number of warnings from clang.
